### DISCUSSION TITLE
Habit Tracker Schema Question

### DISCUSSION THREAD

**Discussion by Mozart409** [prisma/prisma#12601] [Q&A] [answered]
Hello,

I want to create a habit tracker where users can define goals and see if they have done it on a day. My preferred frontend layout would be something like this.

```
           Mo | Tu | We | Th | Fr | Sa | So
Sport      yes| no | yes | no | no | no | yes
```
```
            Mo | Tu | We | Th | Fr | Sa | So
Coffee      yes| no | yes | no | no | no | yes
```

> My current schema

```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id    String @id @default(cuid())
  email String @unique

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  password Password?
  notes    Note[]
  Habit    Habit[]

}

model Password {
  hash String

  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  userId String @unique
}

model Note {
  id    String @id @default(cuid())
  title String
  body  String

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  userId String
}

model Habit {
  id        String      @id @default(cuid())
  label     String      @unique
  HabitItem HabitItem[]

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  user   User   @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: Cascade)
  userId String
}

model HabitItem {
  id String @id @default(cuid())

  completed Boolean @default(false)

  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  Habit   Habit?  @relation(fields: [habitId], references: [id])
  habitId String?

}


```

## Questions

How do I create NON completed HabitItems, when the user does not use the app. 

Habit "Sport"                       
1. User adds a habitItem with completed true on Monday
2. Does not use the app on tu, we, th
3. User adds a new entry on Friday with completed true

How do I automatically add all those NON completed HabitItems for the Habit "Sport"?

**Comment by nurul3101** [comment, marked as answer]
Hey @Mozart409 👋 
Do you want to mark habits as non completed at the end of each day?
Or do you want habits to be marked as non completed only when a user marks a habit completed on a particular day and all the intermediate days get marked as non completed?

Mark habit as non completed at the end of the day if its mark completed could be achieved by using a `cron` which would mark all habits as non completed for all users who haven't already completed.

For the second case, you would need to find the last day on which the habit was marked as completed and then find all habits between the current day and the last day of habit completion and mark all those days in between as non completed.

**Reply by Mozart409** [reply]
Thank you. 👍🏻

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
